### ISSUE TITLE
L2 norm for fp16 goes to nan, needs to be fixed

### ISSUE THREAD

**Issue by rahul003** [apache/incubator-mxnet#11403]
When .norm() method is called on a float16 array it tries to compute the norm in fp16, which causes the squared sum to go out of range and causes nan as output. Large normalizations should be kept in float32, similar to how it's done for softmax and batchnorm. Ideally this should be done by accumulating the squared sum into a float32 buffer. 
This problem might affect the L2 norm operator as well. Haven't verified that.

**Comment by anirudhacharya** [comment]
Yes, I have encountered this bug. Currently ``norm`` operator does not have tests for gradient or symbolic_backward and for different precision formats like fp16 and fp32. There is just one check in ``test_ndarray.py``

**Comment by leezu** [comment]
How about using a numerically stable L2 norm algorithm by scaling the elements? Ie. [BLAS nrm2](http://www.netlib.org/blas/snrm2.f)

**Comment by rahul003** [comment]
That sounds good too

**Comment by leezu** [comment]
@rahul003 would this fix your issue https://github.com/apache/incubator-mxnet/pull/11573 ?

**Comment by haojin2** [comment]
@leezu I can verify the fix through a simple example:
```
>>> a = mx.nd.array([256, 256, 256, 256, 256], dtype=np.float16)
>>> mx.nd.norm(a)
[ 572.5]
<NDArray 1 @cpu(0)>
```
@rahul003 Do you think leo's fix would solve the issue?

**Comment by rahul003** [comment]
Yeah, I think we can close this.

**Comment by haojin2** [comment]
@leezu Thanks for the fix!
